File: src/main/java/org/cerion/stocklist/web/json/Json.java
package org.cerion.stocklist.web.json;

public abstract class Json {

    protected int parsedLength;

    public static Json parse(String json) {
        return parse(json, 0);
    }

    static Json parse(String json, int start) {
        int i = getFirstNonWhitespacePosition(json, start);

        String debug = json.substring(i);

        char c = json.charAt(i);
        if (c == '[') {
            return new JsonArray(json, i);
        } else if (c == '{') {
            return new JsonObject(json, i);
        } else if (c == '"') {
            return new JsonString(json ,i);
        }

        return null;
    }

    static int getFirstNonWhitespacePosition(String json, int start) {
        int i = start;
        while(i < json.length()) {
            char c = json.charAt(i);
            if (!isWhiteSpace(c))
                break;

            i++;
        }

        return i;
    }

    static boolean isWhiteSpace(char c) {
        return c == ' ' || c == '\r' || c == '\n';
    }
}

File: src/main/java/org/cerion/stocklist/web/json/Json.kt
package org.cerion.stocklist.web.json

abstract class Json {

    var parsedLength: Int = 0

    companion object {

        fun parse(json: String): Json? {
            return parse(json, 0)
        }

        internal fun parse(json: String, start: Int): Json? {
            val i = getFirstNonWhitespacePosition(json, start)

            //val debug = json.substring(i)

            val c = json[i]
            if (c == '[') {
                return JsonArray(json, i)
            } else if (c == '{') {
                return JsonObject(json, i)
            } else if (c == '"') {
                return JsonString(json, i)
            }

            return null
        }

        internal fun getFirstNonWhitespacePosition(json: String, start: Int): Int {
            var i = start
            while (i < json.length) {
                val c = json[i]
                if (!isWhiteSpace(c))
                    break

                i++
            }

            return i
        }

        internal fun isWhiteSpace(c: Char): Boolean {
            return c == ' ' || c == '\r' || c == '\n'
        }
    }
}
